src/ol/render/canvas/style.js

```diff
@@ -109,6 +117,14 @@ export function flatStylesToStyleFunction(flatStyles) {
   return function (feature, resolution) {
     evaluationContext.properties = feature.getPropertiesInternal();
     evaluationContext.resolution = resolution;
+    if (parsingContext.featureId) {
+      const id = feature.getId();
+      if (id !== undefined) {
+        evaluationContext.featureId = id;
+      } else {
+        evaluationContext.featureId = null;
+      }
+    }
```

Comment:
I'll follow up with a change that reduces the duplication here and above (and increases the test coverage here - currently only covered in rendering tests).